### PR TITLE
feat: wrap DictionaryFactory in a class

### LINKED ISSUE #33: Add more control on how dictionaries are loaded to memory
Hi @adbar ,

And thanks for the awesome library!

I'm currently wrapping this into a REST API that is called by some microservices.
These services can call to lemmatize words in pretty much any supported language (only a single language at a time).

With the current implementation of simplemma, every time that a request is received for a new language, the previously loaded dictionaries are unloaded:
https://github.com/adbar/simplemma/blob/fd93714b84120f6ebdb53aab2b8955f7a5f2a764/simplemma/simplemma.py#L260-L268

For example:
* Request in English - English dictionary is loaded & lemmatization is performed and cached
* Request in English - English dictionary is already loaded so lemmatization is performed and cached
* Request in German - German dictionary is loaded (and English dictionary offloaded and lemmatization cache cleared) & lemmatization is performed and cached
* Request in English - English dictionary is loaded (and German dictionary offloaded and lemmatization cache cleared) & lemmatization is performed and cached

Imagine that I get 90% request in English and 10% in other languages.
I probably want to keep English loaded at all times and just load the others when needed.

An alternative could be to load all languages when the application start, just as you do when running the libary as a script:
https://github.com/adbar/simplemma/blob/fd93714b84120f6ebdb53aab2b8955f7a5f2a764/simplemma/simplemma.py#L568-L571
However, I'm not sure of how much memory is needed for that and that might be a waste for languages that are rarely used.

So, I think that it would be great to give users more control regarding this.
I can think of a couple of options:
* Give the user methods to load/offload dictionaries from memory
* Have a `get_lang` method with the `lru_cache`

Also, when using the `lru_cache`, I think that the user should be able to define how much memory to allocate to caching.
Currently, there are 2 functions being cached and both have a hardcoded value of 64KB.
Depending on my use case I might want a larger cache, for example.

This would be easy to achieve if simplemma expose a class that must be instantiated before calling `lemmatize` and such methods. In this initialization, the user could set up the cache limits, the dictionaries to preload, the caching strategy for dictionaries, etc.

Wdyt?

### PR DESCRIPTION
**Pull request by juanjoDiaz**
Supersedes #46 and closes #33

At the end, we are simply opting for passing these are additional argument to the functions.

### CODE REVIEW

**Comment by juanjoDiaz** [comment]
With this, #33 is not needed since the LRU cache already has the dictionaries cached in-memory.
So, when calling the loading function again it will just take it from there.

File: simplemma/langdetect.py [hunk 1]
```diff
@@ -30,15 +30,21 @@ def prepare_text(text: str, splitting_regex: Pattern[str] = SPLIT_INPUT) -> List
     return [item[0] for item in counter.most_common(1000)]
 
 
-def in_target_language(text: str, lang: Optional[Tuple[str]] = None) -> float:
+def in_target_language(
+    text: str,
+    lang: Optional[Tuple[str]] = None,
+    dictionaryFactory: DictionaryFactory = DictionaryFactory(),
```
Comment (adbar) [inline simplemma/langdetect.py:36]: "Names should be lowercase, with words separated by underscores as necessary to improve readability" as of PEP 8, so _dictionary_factory_ would be more pythonic.

**Comment by adbar** [comment]
Thanks, the PR looks good!

Regarding the contents we could maybe add a function to unload the superfluous dictionaries in case the user asks for a subset of what's already in memory. Say people load a lot of dicts and then wish to load a list larger than cache limit, e.g. `("ast", "de", "en", "es", "fr", "it", "nl", "pt")` → `("es", "fr", "it")`
It would then be beneficial to go through an unloading phase instead of re-loading every language, what do you think?

 I also have a small comment on style (see above).

File: simplemma/dictionary_factory.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    filename = f"data/{langcode}.plzma"
+    filepath = str(Path(__file__).parent / filename)
+    with lzma.open(filepath, "rb") as filehandle:
+        pickled_dict = pickle.load(filehandle)
+        assert isinstance(pickled_dict, dict)
+        return pickled_dict
+
+
+class DictionaryFactory:
+    def __init__(self, cache_max_size: Optional[int] = 5):
```
Comment (adbar) [inline simplemma/dictionary_factory.py:36]: A power of two for cache size could be better, I suggest 8.
https://stackoverflow.com/questions/61536704/why-does-python-lru-cache-performs-best-when-maxsize-is-a-power-of-two

Comment (juanjoDiaz) [reply simplemma/dictionary_factory.py]: Solved!

File: simplemma/langdetect.py (hunk 1 above)
Comment (juanjoDiaz) [reply simplemma/langdetect.py]: Solved!

**Comment by juanjoDiaz** [comment]
> Regarding the contents we could maybe add a function to unload the superfluous dictionaries in case the user asks for a subset of what's already in memory. Say people want to load a lot of dicts (larger than cache limit), e.g. ("ast", "de", "en", "es", "fr", "it", "nl", "pt") → ("es", "fr", "it")
It would then be beneficial to go through an unloading phase instead of re-loading every language, what do you think?

The problem with that is that it screws up the LRU.

Let's say that your cache is of size 3.
```python
dictionary_factory.get_dictionaries(("en", "de")) // en & de in cache
dictionary_factory.get_dictionaries(("en", "fr")) // en, de & fr in cache (en already in memory)
dictionary_factory.get_dictionaries(("en", "cy")) // de, fr & cy in cache (en already in memory)
dictionary_factory.get_dictionaries(("de", "cy")) // de, fr & cy in cache (en is unloaded)
dictionary_factory.get_dictionaries(("en")) // cy & en in cache (en must be loaded from disk despite having been called so recently)
```

With the simple LRU
```python
dictionary_factory.get_dictionaries(("en", "de")) // en & de in cache
dictionary_factory.get_dictionaries(("en", "fr")) // de, en & fr in cache
dictionary_factory.get_dictionaries(("en", "cy")) // fr, en, & cy in cache
dictionary_factory.get_dictionaries(("de", "cy")) // en, de & cy in cache
dictionary_factory.get_dictionaries(("en")) // en, de & cy in cache (en is already in cache)
```

I've considering implementing a "most often called" cache as an alternative to the LRU cache.

I also have implemented already (didn't include in this PR to keep it smaller) functionality to "preload" a set of languages so they are kept always in memory regardless of the LRU cache rotation.

The options are endless and that's the beauty.
As long as the class implements `def get_dictionaries(self, langs: Optional[Union[str, Tuple[str]]]) -> Dict[str, Dict[str, str]]:`  it will work as a dictionary cache so people can implement whatever strategy they want 🙂

File: simplemma/dictionary_factory.py (hunk 2 above)
Comment (juanjoDiaz) [reply simplemma/dictionary_factory.py]: Changed to 4 but changed now to 8

**Comment by adbar** [comment]
Then it's fine for now, thanks for the changes. The PR looks ready, I'll have a look at it again on Monday and eventually merge it.

**Pull request merged**